File: backend/app/schemas/job.py
"""Pydantic schemas for Industry jobs.

Field names and constraints match database/migrations/019_jobs.sql exactly
(`jobs` + `job_skills`). Validation here mirrors that migration's CHECK
constraints so a bad value comes back as a friendly 422 instead of a raw
database error -- the database stays authoritative.

`jobs` differs from `internships`: it has `employment_type` and
`experience_min_years`, a salary *range* (`salary_min` / `salary_max` /
`salary_currency`), and no `duration_months` / `start_date` /
`stipend_*`.

Ownership (`industry_id`) and lifecycle (`status`) are never accepted from
the client -- `industry_id` is always the authenticated caller, and
`status` only changes through the explicit publish/close/archive
endpoints. `extra="forbid"` on the write models rejects any attempt to
smuggle either one in.
"""

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

JobStatus = Literal["DRAFT", "PUBLISHED", "CLOSED", "ARCHIVED"]
WorkMode = Literal["ONSITE", "REMOTE", "HYBRID"]
EmploymentType = Literal["FULL_TIME", "PART_TIME", "CONTRACT"]
RequiredLevel = Literal["Beginner", "Intermediate", "Advanced", "Expert"]
SkillImportance = Literal["CORE", "IMPORTANT", "OPTIONAL"]

JOB_STATUSES: tuple[str, ...] = ("DRAFT", "PUBLISHED", "CLOSED", "ARCHIVED")
WORK_MODES: tuple[str, ...] = ("ONSITE", "REMOTE", "HYBRID")
EMPLOYMENT_TYPES: tuple[str, ...] = ("FULL_TIME", "PART_TIME", "CONTRACT")


def _blank_strings_to_none(data: object) -> object:
    if not isinstance(data, dict):
        return data
    cleaned: dict = {}
    for key, value in data.items():
        if isinstance(value, str):
            value = value.strip() or None
        cleaned[key] = value
    return cleaned


# ---- skills (job_skills has the same columns as internship_skills) ----


class JobSkillInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    skill_id: UUID
    required_level: RequiredLevel
    importance: SkillImportance = "IMPORTANT"


class JobSkillResponse(BaseModel):
    skill_id: str
    skill_name: str
    category_name: str | None = None
    required_level: str
    importance: str


# ---- jobs ----


class _JobEditableFields(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, min_length=1, max_length=10_000)
    location: str | None = Field(default=None, max_length=200)
    work_mode: WorkMode | None = None
    employment_type: EmploymentType | None = None
    salary_min: float | None = Field(default=None, ge=0)
    salary_max: float | None = Field(default=None, ge=0)
    salary_currency: str | None = Field(default=None, min_length=1, max_length=8)
    experience_min_years: float | None = Field(default=None, ge=0, le=99.9)
    openings: int | None = Field(default=None, ge=1)
    eligibility_criteria: str | None = Field(default=None, max_length=5_000)
    application_deadline: date | None = None

    @model_validator(mode="before")
    @classmethod
    def _normalise_blanks(cls, data: object) -> object:
        return _blank_strings_to_none(data)

    @model_validator(mode="after")
    def _check_salary_range(self) -> "_JobEditableFields":
        if (
            self.salary_min is not None
            and self.salary_max is not None
            and self.salary_max < self.salary_min
        ):
            raise ValueError("Maximum salary can't be lower than the minimum salary.")
        return self


class JobCreate(_JobEditableFields):
    """POST /api/v1/jobs. Always created as DRAFT -- there is no `status`
    field here, and no `industry_id`."""

    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=10_000)
    skills: list[JobSkillInput] = Field(default_factory=list, max_length=50)


class JobUpdate(_JobEditableFields):
    """PUT /api/v1/jobs/{id}. Partial: only the fields actually sent are
    changed. `skills` omitted = leave the skill list alone; `skills`
    present = replace it wholesale. `status` cannot be changed here."""

    model_config = ConfigDict(extra="forbid")

    skills: list[JobSkillInput] | None = Field(default=None, max_length=50)


class JobResponse(BaseModel):
    id: str
    industry_id: str
    title: str
    description: str
    location: str | None = None
    work_mode: str | None = None
    employment_type: str | None = None
    salary_min: float | None = None
    salary_max: float | None = None
    salary_currency: str | None = None
    experience_min_years: float | None = None
    openings: int | None = None
    eligibility_criteria: str | None = None
    application_deadline: str | None = None
    status: str
    created_at: str | None = None
    updated_at: str | None = None
    skills: list[JobSkillResponse] = Field(default_factory=list)


class JobListResponse(BaseModel):
    jobs: list[JobResponse]

File: frontend/types/job.ts
// Mirrors the `jobs` + `job_skills` tables
// (database/migrations/019_jobs.sql) and backend/app/schemas/job.py.
// Keep all three in sync.
//
// `jobs` differs from `internships`: it has `employment_type` and
// `experience_min_years`, a salary *range* (salary_min / salary_max /
// salary_currency), and no `duration_months` / `start_date` / stipend.

import type { SkillRequirement, SkillRequirementInput } from "@/types/skill-requirement";

export {
  REQUIRED_LEVELS,
  SKILL_IMPORTANCES,
  SKILL_IMPORTANCE_LABELS,
} from "@/types/skill-requirement";
export type { RequiredLevel, SkillImportance } from "@/types/skill-requirement";

export const JOB_STATUSES = ["DRAFT", "PUBLISHED", "CLOSED", "ARCHIVED"] as const;
export type JobStatus = (typeof JOB_STATUSES)[number];

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  DRAFT: "Draft",
  PUBLISHED: "Published",
  CLOSED: "Closed",
  ARCHIVED: "Archived",
};

export const WORK_MODES = ["ONSITE", "REMOTE", "HYBRID"] as const;
export type WorkMode = (typeof WORK_MODES)[number];

export const WORK_MODE_LABELS: Record<WorkMode, string> = {
  ONSITE: "On-site",
  REMOTE: "Remote",
  HYBRID: "Hybrid",
};

export const EMPLOYMENT_TYPES = ["FULL_TIME", "PART_TIME", "CONTRACT"] as const;
export type EmploymentType = (typeof EMPLOYMENT_TYPES)[number];

export const EMPLOYMENT_TYPE_LABELS: Record<EmploymentType, string> = {
  FULL_TIME: "Full-time",
  PART_TIME: "Part-time",
  CONTRACT: "Contract",
};

/** A skill requirement as returned on a job (catalog data joined in). */
export type JobSkill = SkillRequirement;

/** A skill requirement as sent to the API on create/update. */
export type JobSkillInput = SkillRequirementInput;

export interface Job {
  id: string;
  industry_id: string;
  title: string;
  description: string;
  location: string | null;
  work_mode: WorkMode | null;
  employment_type: EmploymentType | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  experience_min_years: number | null;
  openings: number | null;
  eligibility_criteria: string | null;
  application_deadline: string | null;
  status: JobStatus;
  created_at: string | null;
  updated_at: string | null;
  skills: JobSkill[];
}

/** POST body. Always created as DRAFT — no `status`, no `industry_id`. */
export interface JobCreate {
  title: string;
  description: string;
  location: string | null;
  work_mode: WorkMode | null;
  employment_type: EmploymentType | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  experience_min_years: number | null;
  openings: number | null;
  eligibility_criteria: string | null;
  application_deadline: string | null;
  skills: JobSkillInput[];
}

/** PUT body — partial. Any field omitted is left unchanged; `skills`
 * omitted leaves the list alone, a list replaces it. */
export type JobUpdate = Partial<JobCreate>;
